src/index.js

```diff
@@ -505,50 +742,62 @@ class Service extends AdapterService {
+  _update (id, data, params = {}) {
+    // NOTE : First fetch the item to update to account for user query
     return this._get(id, params)
-      .then(oldData => {
-        let newObject = {};
+      .then(() => {
+        // NOTE: Next, fetch table metadata so
+        // that we can fill any existing keys that the
+        // client isn't updating with null;
+        return this.Model.fetchTableMetadata()
+          .then(async meta => {
+            let newObject = Object.assign({}, data);
+            let transaction = null;
+
+            const allowedUpsert = this.mergeRelations(this.allowedUpsert, params.mergeAllowUpsert);
+
+            if (allowedUpsert) {
+              // Ensure the object we fetched is the one we update
+              this._checkUpsertId(id, newObject);
+              // Create transaction if needed
+              transaction = await this._createTransaction(params);
+            }
 
-        for (const key of Object.keys(oldData)) {
-          if (data[key] === undefined) {
-            newObject[key] = null;
-          } else {
-            newObject[key] = data[key];
-          }
-        }
+            for (const key of meta.columns) {
+              if (newObject[key] === undefined) {
+                newObject[key] = null;
+              }
+            }
 
-        const allowedUpsert = this.mergeRelations(this.allowedUpsert, params.mergeAllowUpsert);
-        if (allowedUpsert) {
-          return this.createQuery(params)
-            .allowGraph(allowedUpsert)
-            .upsertGraphAndFetch(newObject, this.upsertGraphOptions);
-        }
+            if (allowedUpsert) {
+              const upsertGraphOptions = { ...this.upsertGraphOptions, ...params.mergeUpsertGraphOptions };
+              return this._createQuery(params)
+                .allowGraph(allowedUpsert)
+                .upsertGraphAndFetch(newObject, upsertGraphOptions).then(this._commitTransaction(transaction), this._rollbackTransaction(transaction));
+            }
 
-        // NOTE (EK): Delete id field so we don't update it
-        if (Array.isArray(this.id)) {
-          for (const idKey of this.id) {
-            delete newObject[idKey];
-          }
-        } else {
-          delete newObject[this.id];
-        }
-        return this.createQuery(params)
-          .where(this.getIdsQuery(id))
-          .update(newObject)
-          .then(() => {
-            // NOTE (EK): Restore the id field so we can return it to the client
+            // NOTE (EK): Delete id field so we don't update it
             if (Array.isArray(this.id)) {
-              newObject = Object.assign({}, newObject, this.getIdsQuery(id));
+              for (const idKey of this.id) {
+                delete newObject[idKey];
+              }
             } else {
-              newObject[this.id] = id;
+              delete newObject[this.id];
             }
-
-            return newObject;
-          });
+            return this._createQuery(params)
```

Comment:
Can you please look into this and see if `this._createQuery` should be `this.createQuery`.